# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/gamepad/ipc/GamepadEventChannelParent.cpp
# Commit: d31313f495aa
# Full Hash: d31313f495aa20866d041d044cca726b570f392c
# Author: Chris Martin <cmartin@mozilla.com>
# Date: 2021-01-05 04:31:31
# Regressor Bug: 1657404
# File Overlap Count: 2
# Description:
#   Bug 1657404 - Add diag asserts to catch potential double IPDL shutdown r=handyman
#   
#   Based on evidence from crash reports in Bug 1682589, there may be a bug where
#   IPDL intermittently calls `ActorDestroy()` twice on GamepadEventChannelParent,
#   leading to a crash.
# ==============================================================================

diff -r 585c85986b92 -r d31313f495aa dom/gamepad/ipc/GamepadEventChannelParent.cpp
--- a/dom/gamepad/ipc/GamepadEventChannelParent.cpp	Mon Jan 04 20:57:43 2021 +0000
+++ b/dom/gamepad/ipc/GamepadEventChannelParent.cpp	Mon Jan 04 20:38:39 2021 +0000
@@ -44,8 +44,8 @@
       .forget();
 }
 
-GamepadEventChannelParent::GamepadEventChannelParent() {
-  AssertIsOnBackgroundThread();
+GamepadEventChannelParent::GamepadEventChannelParent() : mIsShutdown{false} {
+  MOZ_DIAGNOSTIC_ASSERT(IsOnBackgroundThread());
 
   mBackgroundEventTarget = GetCurrentEventTarget();
 
@@ -57,7 +57,10 @@
 }
 
 void GamepadEventChannelParent::ActorDestroy(ActorDestroyReason aWhy) {
-  AssertIsOnBackgroundThread();
+  MOZ_DIAGNOSTIC_ASSERT(IsOnBackgroundThread());
+  MOZ_DIAGNOSTIC_ASSERT(!mIsShutdown);
+
+  mIsShutdown = true;
 
   RefPtr<GamepadPlatformService> service =
       GamepadPlatformService::GetParentService();